### PR TITLE
Update middleware upgrade guide w/ more rationale

### PR DESCRIPTION
**Pull request by ethomson**
The API changes for GA are general improvements.

Our middleware upgrade guide is very helpful about describing the changes
that were made and how to react to them; include some additional "why" to
rationalize these changes.

## Bug

- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have helpful link attached, see `contributing.md`

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [ ] Make sure the linting passes by running `pnpm lint`
- [ ] The examples guidelines are followed from [our contributing doc](https://github.com/vercel/next.js/blob/canary/contributing.md#adding-examples)

### CODE REVIEW

File: errors/middleware-upgrade-guide.md [hunk 1]
```diff
@@ -2,13 +2,15 @@
...
 
-This upgrade guide will help you understand the changes and how to migrate your existing Middleware to the new API. The guide is for Next.js developers who:
+This upgrade guide will help you understand the changes, why they were made, and how to migrate your existing Middleware to the new API. The guide is for Next.js developers who:
 
 - Currently use the beta Next.js Middleware features
 - Choose to upgrade to the next stable version of Next.js (`v12.2`)
 
 You can start upgrading your Middleware usage today with the latest canary release (`npm i next@canary`).
 
+**Note**: these changes described in this guide are included in Next.js 12.2. You can keep your current site structure, including nested middleware until you move to 12.2 (or a canary build of Next.js).
```
Comment (leerob) [inline errors/middleware-upgrade-guide.md:12]:
```suggestion
> **Note**: These changes described in this guide are included in Next.js `12.2`. You can keep your current site structure, including nested Middleware, until you move to `12.2` (or a `canary` build of Next.js).
```

File: errors/middleware-upgrade-guide.md [hunk 2]
```diff
@@ -34,13 +36,20 @@ If you're using Next.js on Vercel, your existing deploys using Middleware will c
...
+Previously, you could create a `_middleware.ts` file under the `pages` directory at any level. Middleware execution was based on the file path where it was created.
 
-- Middleware in `pages/dashboard/_middleware.ts`
-- Middleware in `pages/dashboard/users/_middleware.ts`
-- A request to `/dashboard/users/*` **would match both.**
+Based on customer feedback, we have replaced this API with a single root Middleware, which provides the following improvements:
 
-Based on customer feedback, we have replaced this API with a single root Middleware.
+- **Faster execution with lower latency**  
+  With nested middleware, a single request could invoke multiple middleware functions. A single middleware means a single function execution, which is more efficient.
```
Comment (leerob) [inline errors/middleware-upgrade-guide.md:44]: Can we put these like:

- **Thing:** description

File: errors/middleware-upgrade-guide.md [hunk 3]
```diff
@@ -90,7 +99,7 @@ export function middleware(request: NextRequest) {
 
 ### Explanation
 
-To help ensure security, we are removing the ability to send response bodies in Middleware. This ensures that Middleware is only used to `rewrite`, `redirect`, or modify the incoming request (e.g. [setting cookies](#cookies-api-revamped)).
+To respect the differences in client-side and server-side navigation, and to help ensure that developers do not build insecure middleware, we are removing the ability to send response bodies in Middleware. This ensures that Middleware is only used to `rewrite`, `redirect`, or modify the incoming request (e.g. [setting cookies](#cookies-api-revamped)).
```
Comment (leerob) [inline errors/middleware-upgrade-guide.md:102]:
```suggestion
To respect the differences in client-side and server-side navigation, and to help ensure that developers do not build insecure Middleware, we are removing the ability to send response bodies in Middleware. This ensures that Middleware is only used to `rewrite`, `redirect`, or modify the incoming request (e.g. [setting cookies](#cookies-api-revamped)).
```

File: errors/middleware-upgrade-guide.md (hunk 2 above)
Comment (ethomson) [reply errors/middleware-upgrade-guide.md]: Done. 👍

File: errors/middleware-upgrade-guide.md [hunk 4]
```diff
@@ -34,13 +36,15 @@ If you're using Next.js on Vercel, your existing deploys using Middleware will c
...
+- **Straighforward, deterministic execution ordering**: with nested middleware, a single request could match multiple middleware functions. For example, a requests to `/dashboard/users/*` would invoke middleware defined in both `/dashboard/users/_middleware.ts` and `/dashboard/_middleware.js`, however the execution order is not straightforward to reason about. Moving to a single, root middleware puts execution order in the hand of the developer, so they can implement the order that makes the most sense for any given project.
+- **Supports Next.js Layouts (RFC)**: moving to a single, root middleware helps support [Layouts (RFC) in Next.js](https://nextjs.org/blog/layouts-rfc). The Layouts RFC proposes to move away from the current `pages` model, which doesn’t support a nested middleware model, but does support a single, root middleware instead.
```
Comment (leerob) [inline errors/middleware-upgrade-guide.md:47]:
```suggestion
- **Faster execution with lower latency**: With nested Middleware, a single request could invoke multiple Middleware functions. A single Middleware means a single function execution, which is more efficient.
- **Less expensive**: Middleware usage is billed per invocation. Using nested Middleware, a single request could invoke multiple Middleware functions, meaning multiple Middleware charges per request. A single Middleware means a single invocation per request and is more cost effective.
- **Middleware can conveniently filter on things besides routes**: With nested Middleware, the Middleware files were located in the `pages` directory and Middleware was executed based on request paths. By moving to a single root Middleware, you can still execute code based on request paths, but you can now more conveniently execute Middleware based on other conditions, like `cookies` or the presence of a request header.
- **Deterministic execution ordering**: With nested Middleware, a single request could match multiple Middleware functions. For example, a request to `/dashboard/users/*` would invoke Middleware defined in both `/dashboard/users/_middleware.ts` and `/dashboard/_middleware.js`. However, the execution order is not easy to reason about. Moving to a single, root Middleware more explicitly defines execution order.
- **Supports Next.js Layouts (RFC)**: Moving to a single, root mMddleware helps support the new [Layouts (RFC) in Next.js](https://nextjs.org/blog/layouts-rfc).
```

File: errors/middleware-upgrade-guide.md [hunk 5]
```diff
@@ -34,13 +36,15 @@ If you're using Next.js on Vercel, your existing deploys using Middleware will c
...
+- **Less expensive**: Middleware usage is billed per invocation. Using nested Middleware, a single request could invoke multiple Middleware functions, meaning multiple Middleware charges per request. A single Middleware means a single invocation per request and is more cost effective.
+- **Middleware can conveniently filter on things besides routes**: With nested Middleware, the Middleware files were located in the `pages` directory and Middleware was executed based on request paths. By moving to a single root Middleware, you can still execute code based on request paths, but you can now more conveniently execute Middleware based on other conditions, like `cookies` or the presence of a request header.
+- **Deterministic execution ordering**: With nested Middleware, a single request could match multiple Middleware functions. For example, a request to `/dashboard/users/*` would invoke Middleware defined in both `/dashboard/users/_middleware.ts` and `/dashboard/_middleware.js`. However, the execution order is not easy to reason about. Moving to a single, root Middleware more explicitly defines execution order.
```
Comment (balazsorban44) [inline errors/middleware-upgrade-guide.md:46]:
```suggestion
- **Deterministic execution ordering**: With nested Middleware, a single request could match multiple Middleware functions. For example, a request to `/dashboard/users/*` would invoke Middleware defined in both `/dashboard/users/_middleware.ts` and `/dashboard/_middleware.js`. However, the execution order is harder to reason about. Moving to a single, root Middleware more explicitly defines execution order.
```

**Review by leerob — DISMISSED** [review]

File: errors/middleware-upgrade-guide.md [hunk 6]
```diff
@@ -34,13 +36,15 @@ If you're using Next.js on Vercel, your existing deploys using Middleware will c
...
+- **Less expensive**: Middleware usage is billed per invocation. Using nested Middleware, a single request could invoke multiple Middleware functions, meaning multiple Middleware charges per request. A single Middleware means a single invocation per request and is more cost effective.
+- **Middleware can conveniently filter on things besides routes**: With nested Middleware, the Middleware files were located in the `pages` directory and Middleware was executed based on request paths. By moving to a single root Middleware, you can still execute code based on request paths, but you can now more conveniently execute Middleware based on other conditions, like `cookies` or the presence of a request header.
+- **Deterministic execution ordering**: With nested Middleware, a single request could match multiple Middleware functions. For example, a request to `/dashboard/users/*` would invoke Middleware defined in both `/dashboard/users/_middleware.ts` and `/dashboard/_middleware.js`. However, the execution order is harder to reason about. Moving to a single, root Middleware more explicitly defines execution order.
```
Comment (balazsorban44) [inline errors/middleware-upgrade-guide.md:46]:
```suggestion
- **Deterministic execution ordering**: With nested Middleware, a single request could match multiple Middleware functions. For example, a request to `/dashboard/users/*` would invoke Middleware defined in both `/dashboard/users/_middleware.ts` and `/dashboard/_middleware.js`. However, the execution order is difficult to reason about. Moving to a single, root Middleware more explicitly defines execution order.
```

https://github.com/vercel/next.js/runs/7082368888?check_suite_focus=true#step:6:251

File: errors/middleware-upgrade-guide.md [hunk 7]
```diff
@@ -2,13 +2,15 @@
...
+This upgrade guide will help you understand the changes, why they were made, and how to migrate your existing Middleware to the new API. The guide is for Next.js developers who:
 
 - Currently use the beta Next.js Middleware features
 - Choose to upgrade to the next stable version of Next.js (`v12.2`)
 
 You can start upgrading your Middleware usage today with the latest canary release (`npm i next@canary`).
 
+> **Note**: These changes described in this guide are included in Next.js `12.2`. You can keep your current site structure, including nested Middleware, until you move to `12.2` (or a `canary` build of Next.js).
+
 If you have ESLint configured, you will need to run `npm i eslint-config-next@canary --save-dev` to upgrade your ESLint configuration to ensure the same version is being used as the Next.js version. You might also need to restart VSCode for the changes to take effect.
```
Comment (ijjk) [inline errors/middleware-upgrade-guide.md:14]:
```suggestion
If you have ESLint configured, you will need to run `npm i eslint-config-next@latest --save-dev` to upgrade your ESLint configuration to ensure the same version is being used as the Next.js version. You might also need to restart VSCode for the changes to take effect.
```

File: errors/middleware-upgrade-guide.md [hunk 8]
```diff
@@ -34,13 +36,15 @@ If you're using Next.js on Vercel, your existing deploys using Middleware will c
...
+- **Middleware can conveniently filter on things besides routes**: With nested Middleware, the Middleware files were located in the `pages` directory and Middleware was executed based on request paths. By moving to a single root Middleware, you can still execute code based on request paths, but you can now more conveniently execute Middleware based on other conditions, like `cookies` or the presence of a request header.
+- **Deterministic execution ordering**: With nested Middleware, a single request could match multiple Middleware functions. For example, a request to `/dashboard/users/*` would invoke Middleware defined in both `/dashboard/users/_middleware.ts` and `/dashboard/_middleware.js`. However, the execution order is difficult to reason about. Moving to a single, root Middleware more explicitly defines execution order.
+- **Supports Next.js Layouts (RFC)**: Moving to a single, root mMddleware helps support the new [Layouts (RFC) in Next.js](https://nextjs.org/blog/layouts-rfc).
```
Comment (ijjk) [inline errors/middleware-upgrade-guide.md:47]:
```suggestion
- **Supports Next.js Layouts (RFC)**: Moving to a single, root Mddleware helps support the new [Layouts (RFC) in Next.js](https://nextjs.org/blog/layouts-rfc).
```

**Review by ijjk — APPROVED** [review]

**Pull request merged**